src/NITTA/LuaFrontend.hs

```diff
@@ -15,460 +11,397 @@ Copyright   : (c) Aleksandr Penskoi, 2019
-            -- `expConstant` return a wrong name, the correct name will be established
-            -- after all statement was processed
-            void $ expConstant name n
-            flushBuffer diff []
-        _ -> error "internal error"
-
--- e.g. @n, d = a / b@, or @n, d = f()@
-processStatement _fn (Assign lexps [rexp])
-    | length lexps > 1 = do
-        let outs = parseLeftExp lexps
-        diff <- renameVarsIfNeeded outs
-        rightExp diff outs rexp
-        flushBuffer diff outs
-
--- e.g. @a = 1@ or @a, b = 1, 2@
-processStatement _fn st@(Assign lexps rexps)
-    | length lexps == length rexps = do
-        let outs = parseLeftExp lexps
-        diff <- renameVarsIfNeeded outs
-        zipWithM_ (rightExp diff) (map (: []) outs) rexps
-        flushBuffer diff outs
-    | otherwise = error $ "assignment mismatch: " ++ show st
--- recursive call of main function
+processStatement _ (Assign lexp [rexp]) = do
+    parseRightExp (map parseLeftExp lexp) rexp
+processStatement startupFunctionName (Assign vars exps) | length vars == length exps = do
+    mapM_ (\(VarName (Name name), expr) -> processStatement startupFunctionName (Assign [VarName (Name (getTempAlias name))] [expr])) $ zip vars exps
+    mapM_ (\(VarName (Name name)) -> addAlias name (getTempAlias name)) vars
+    where
+        getTempAlias name = name <> "&"
+--startup function recursive call
 processStatement fn (FunCall (NormalFunCall (PEVar (VarName (Name fName))) (Args args)))
-    | fn == fName =
-        do
-            AlgBuilder{algItems} <- get
-            let algIn = reverse $ filter (\case InputVar{} -> True; _ -> False) algItems
-            mapM_ (uncurry f) $ zip algIn args
+    | fn == fName = do
+        LuaAlgBuilder{algStartupArgs} <- get
+        let startupVarsNames = map ((\(Just x) -> x) . (`HM.lookup` algStartupArgs)) [0 .. (HM.size algStartupArgs)]
+        let startupVarsVersions = map (\x -> LuaValueInstance{luaValueInstanceName = fst x, luaValueInstanceAssignCount = 0, luaValueInstanceIsConstant = False}) startupVarsNames
+        mapM_ parseStartupArg $ zip3 args startupVarsVersions (map (readText . snd) startupVarsNames)
     where
-        f InputVar{iX, iVar} rexp = do
-            i <- expArg [] rexp
-            let loop = Function{fName = "loop", fIn = [i], fOut = [iVar], fValues = [iX], fInt = []}
-            modify'_ $ \alg@AlgBuilder{algItems} -> alg{algItems = loop : algItems}
-        f _ _ = undefined
+        parseStartupArg (arg, valueVersion, index) = do
+            varName <- parseExpArg (T.pack "loop") arg
+            luaAlgBuilder@LuaAlgBuilder{algGraph} <- get
+            put luaAlgBuilder{algGraph = LuaStatement{fIn = [varName], fOut = [valueVersion], fValues = [index], fName = "loop", fInt = []} : algGraph}
+processStatement _ (FunCall (NormalFunCall (PEVar (VarName (Name fName))) (Args args))) = do
+    fIn <- mapM (parseExpArg "tmp") args
+    addFunction (fromString $ T.unpack fName) fIn [fromString ""]
```

Comment:
Add fromText to Utility

https://github.com/ryukzak/nitta/blob/master/src/NITTA/Utils/Base.hs

Reply:
Here https://github.com/ryukzak/nitta/pull/183, I add auto checks by hlint.